### PR TITLE
Calendar calendar state context set value can not set null value

### LINKED ISSUE #6006: Calendar's controlled value prop (in onChange) and setValue() from CalendarStateContext can't be set to null
### Provide a general summary of the issue here

I get the `setValue` method from `CalendarStateContext` and try to set `null` as the value for `setValue()` but an error immediately occurs and nothing is set. Prop of `value` and `onChange` and `setValue(null)` with the wrong type and TS swears at `null`. 

### 🤔 Expected Behavior?

Calendar's controlled `value` prop and `state.setValue()` from `CalendarStateContext` and in `onChange` prop can be set to `null`

### 😯 Current Behavior

Calendar controlled `value` and `setValue()` from `CalendarStateContext` can't be set to `null`, judging by `value?:T`, but for some reason it allows you to set `null`.
And the same goes for method `setValue()` from `CalendarStateContext`. You cannot set `null` in `setValue()`

### 💁 Possible Solution

1. Add to the interface `ValueBase<T, C = T>`:

`value?: T | null`,
`onChange?: (value: C (and null))`

```
export interface ValueBase<T, C = T> {
  /** The current value (controlled). */
  value?: T,
  /** The default value (uncontrolled). */
  default Value?: T,
  /** Handler that is called when the value changes. */
  onChange?: (value: C) => void
}
```
 
2. Add to the interface `CalendarState extends CalendarStateBase`:

`setValue(value: CalendarDate | null): void;`
```
export interface CalendarState extends CalendarStateBase {
    /** The currently selected date. */
    readonly value: CalendarDate;
    /** Sets the currently selected date. */
    setValue(value: CalendarDate): void;
}
```

3. Give the option (fix) to set `null` so that you can, for example, reset the date in the calendar.

### 🔦 Context

_No response_

### 🖥️ Steps to Reproduce

[Here you can see an example](https://codesandbox.io/p/sandbox/calendar-from-react-aria-components-67xhm4?file=%2Fsrc%2FApp.tsx%3A21%2C45). You can see that `onChange` and `setValue(null)` with the wrong type and TS swears at null. And if you press the "Clear" button and try to set `state.setValue(null);` and the value in value is not reset

### Version

1.1.1 react-aria-components

### What browsers are you seeing the problem on?

Chrome

### If other, please specify.

_No response_

### What operating system are you using?

MacOS 13.3

### 🧢 Your Company/Team

_No response_

### 🕷 Tracking Issue

_No response_

### PR DESCRIPTION
**Pull request by psywalker**
Closes #6006

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues/6006).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

## 📝 Test Instructions:

1. Open the calendar page where the Footer component is inserted ([Story link](http://localhost:9003/?path=/story/react-aria-components--calendar-reset-value&providerSwitcher-express=false&strict=true))
2. Select an arbitrary date in the calendar.
3. Click the "Reset value" button in the Footer component.
4. Make sure that after clicking the button, the selected date is reset, and no date is displayed in the calendar as selected.
5. Check that pressing the "Reset value" button again after selecting a new date resets the selection again.

https://github.com/adobe/react-spectrum/assets/1297755/8b9f63cc-e333-41d8-81eb-119938665473

### CODE REVIEW

File: packages/react-aria-components/stories/Calendar.stories.tsx [hunk 1]
```diff
@@ -10,13 +10,32 @@
...
+        onPress={() => {
+          // reset value 
+          setValue(null);
+        }}>
+        Reset value
+      </Button>
+    </div>
+  );
+}
+
```
Comment (LFDanLu) [inline packages/react-aria-components/stories/Calendar.stories.tsx:38]: Ah looks like typescript is complaining about the potential that state is null here if the CalendarStateContext Provider doesn't exist above it. You can prob fix it like 
```suggestion
function Footer() {
  const state = useContext(CalendarStateContext);
  const setValue = state?.setValue;

  return (
    <div>
      <Button
        slot={null}
        className="reset-button"
        onPress={() => {
          // reset value
          setValue?.(null);
        }}>
        Reset value
      </Button>
    </div>
  );
}
```

or doing something like 
```
  const state = useContext(CalendarStateContext) || {} as CalendarState;
  const {setValue} = state;
```

Comment (psywalker) [reply packages/react-aria-components/stories/Calendar.stories.tsx]: Hello!

Thank you very much for the solution provided. I did as you said. I fixed everything.

But now the problem is that when checking `ci/circleci:test-canary`, the following error occurs
[https://app.circleci.com/pipelines/github/adobe/react-spectrum/19333/workflows/f8dae52f-81af-4112-bf3b-e22f2f376a76/jobs/281883](https://app.circleci.com/pipelines/github/adobe/react-spectrum/19333/workflows/f8dae52f-81af-4112-bf3b-e22f2f376a76/jobs/281883/parallel-runs/2/steps/2-103)

But the fact is that this error does not apply to the test file that I wrote. For my part, I have corrected everything.
Please take a look when you have time.

PR is ready for merge

Comment (LFDanLu) [reply packages/react-aria-components/stories/Calendar.stories.tsx]: Might be https://github.com/adobe/react-spectrum/pull/6059, I've pulled in main to see if it gets fixed

**Comment by LFDanLu** [comment]
GET_BUILD

File: packages/react-aria-components/stories/Calendar.stories.tsx (hunk 1 above)
Comment (psywalker) [reply packages/react-aria-components/stories/Calendar.stories.tsx]: Thank you! So everything is fine and you can already make merges in the main? :)

Comment (LFDanLu) [reply packages/react-aria-components/stories/Calendar.stories.tsx]: yep, I'll review the changes today and see if I can get another pair of eyes on it as well

**Review by LFDanLu — APPROVED** [review]
LGTM, thanks for the PR!

File: packages/react-aria-components/docs/Calendar.mdx [hunk 2]
```diff
@@ -859,6 +859,55 @@ function CalendarValue() {
 </Calendar>
 ```
 
+#### Reset value
+
+This example shows the `Footer` component, which is placed inside the `Calendar`. In this component, we use `CalendarStateContext` to get the `setValue` method from the `CalendarState` object. Next, by clicking on the button, we set the `setValue` to `null` to reset the `Calendar` `value`.
```
Comment (snowystinger) [inline packages/react-aria-components/docs/Calendar.mdx:864]:
```suggestion
This example shows a custom `Footer` component that can be placed inside the `Calendar.` In this component, we use `CalendarStateContext` to get the `setValue` method from the `CalendarState` object. Next, by clicking on the button, we set the `setValue` to `null` to reset the `Calendar` `value`.
```

File: packages/react-aria-components/test/Calendar.test.js [hunk 3]
```diff
@@ -284,4 +284,62 @@ describe('Calendar', () => {
...
+        <Footer />
+      </Calendar>
+    );
+    let grid = getByRole('application');
+    expect(grid).toHaveAttribute('class', 'grid');
+
+    let cell = within(grid).getAllByRole('button')[7];
+    expect(cell).toBeInTheDocument();
+
+    fireEvent.click(cell);
```
Comment (snowystinger) [inline packages/react-aria-components/test/Calendar.test.js:334]:
```suggestion
    await user.click(cell);
```

File: packages/react-aria-components/test/Calendar.test.js [hunk 4]
```diff
@@ -284,4 +284,62 @@ describe('Calendar', () => {
...
+    expect(cell).toHaveAttribute('data-selected', 'true');
+    expect(cell).toHaveClass('selected');
+
+    const resetButton = grid.querySelector('.reset-button');
+    expect(resetButton).toBeInTheDocument();
+
+    fireEvent.click(resetButton);
```
Comment (snowystinger) [inline packages/react-aria-components/test/Calendar.test.js:341]:
```suggestion
    await user.click(resetButton);
```

**Comment by snowystinger** [comment]
Looks really good, just a couple small things and then I think it's good to go :)

**Comment by dannify** [comment]
hey @psywalker 
It looks like there are some left over review suggestions that need to be incorporated into the PR. After those are cleaned up, we can pick this up again. Thanks!

**Review by snowystinger — APPROVED** [review]

**Review by LFDanLu — APPROVED** [review]

**Comment by andredewaard** [comment]
I dont see this change in the main branch yet? 
https://github.com/psywalker/react-spectrum/blob/a8ec9667a2b70b88b45e329bf44918f0f93d54f9/packages/%40react-stately/calendar/src/types.ts#L104

**Comment by andredewaard** [comment]
Can we also add this to the date range picker?

**Comment by LFDanLu** [comment]
@andredewaard mind expanding a bit on the ask here? DateRangePicker should already support something like this via its own context: https://react-spectrum.adobe.com/react-aria/DateRangePicker.html#state

**Pull request merged**